src/utils.js

```diff
@@ -200,32 +220,183 @@ export const getAgeData = (birthMonth, birthYear, age, ageMonths) => {
  */
 export const getGrade = (inputGrade, gradeMin = 0, gradeMax = 12) => {
-  const parsedGrade = Number(inputGrade)
+  const parsedGrade = Number(inputGrade);
   let grade;
 
-  if (isNaN(parsedGrade)) {
+  if (Number.isNaN(parsedGrade)) {
     // Assume grade is K, TK, or PK
-    grade = gradeMin
+    grade = gradeMin;
   } else if (parsedGrade < gradeMin) {
     grade = gradeMin;
   } else if (parsedGrade > gradeMax) {
     grade = gradeMax;
   } else {
     // grade is within range and is a number
-    grade = parsedGrade
+    grade = parsedGrade;
   }
 
   return grade;
 };
- 
 
+/**
+ * Return grade after min/max filtering and accounting for string values.
+ *
+ * @const median
+ *
+ * @param {array} array - Array of response times.
+ *
+ * @returns {number} median of array of response times
+ */
+export const median = (array) => {
+  array.sort((a, b) => b - a);
+  const { length } = array;
+  if (length % 2 === 0) {
+    return (array[length / 2] + array[length / 2 - 1]) / 2;
+  }
+  return array[Math.floor(length / 2)];
+};
+
+/**
+ * Returns a function that evaluates the reliability of a run based on the following criteria:
+ *
+ * @param {number} RESPONSE_TIME_LOW_THRESHOLD - The minimum acceptable response time threshold.
+ * @param {number} RESPONSE_TIME_HIGH_THRESHOLD - The maximum acceptable response time threshold.
+ * @param {number} RESPONSE_SIMILARITY_THRESHOLD - The maximum acceptable response similarity threshold.
+ * @param {number} ACCURACY_THRESHOLD - The minimum acceptable accuracy threshold.
+ * @param {array} ignoredReliabilityFlags - An array of flags that should be ignored when evaluating reliability.
+ * @returns {function} baseValidityEvaluator - A function that evaluates the reliability of a run.
+ */
+export function CreateEvaluateValidity({
+  RESPONSE_TIME_LOW_THRESHOLD = 400,
+  RESPONSE_TIME_HIGH_THRESHOLD = 10000,
+  RESPONSE_SIMILARITY_THRESHOLD = 0.8,
+  ACCURACY_THRESHOLD = 0.2,
+  ignoredReliabilityFlags = [],
+}) {
+  return function baseEvaluateValidity ({ responseTimes, responses, correct }) {
+    const flags = [];
+
+    // verifies if responseTimes lie above or below a threshold
+    if (median(responseTimes) <= RESPONSE_TIME_LOW_THRESHOLD) {
+      flags.push('responseTimeTooFast');
+    }
+
+    if (median(responseTimes) >= RESPONSE_TIME_HIGH_THRESHOLD) {
+      flags.push('responseTimeTooSlow');
+    }
+
+    // Calculates the response with the highest frequency and return the frequency
+    const maxIdenticalResponse = Math.max(
+      ...Object.values(
+        responses.reduce((acc, val) => {
+          if (!acc[val]) {
+            acc[val] = 0;
+          }
+          acc[val] += 1;
+          return acc;
+        }, {}),
+      ),
+    );
+
+    const similarity = maxIdenticalResponse / responses.length;
+
+    // Calculate response similarity based on maxIdenticalResponse
+    if (similarity >= RESPONSE_SIMILARITY_THRESHOLD) {
+      flags.push('responsesTooSimilar');
+    }
```

Comment:
for this part, 

i am thinking a bit differently to determine "responsesTooSimilar'. instead of tracking max one type of responses / total responses, we want to track a sequence of responses and see if they are not spamming. For example, for every consecutive 10 responses, if their responses are ALL same, we should give them "responsesTooSimilar" flag. then response_similarity_threshold should be treated as "how many responses we should count as a sequence, e.g., 10" to track spamming. happy to elaborate more over slack.

Reply:
@richford for comments

Reply:
This makes sense, so perhaps if RESPONSE_SIMILARITY_THRESHOLD = n, it should track if the last n responses to see if they are identical?

Reply:
That makes sense, let me update the function.